Blog post AI-Assisted IaC with Pulumi's Model Context Protocol Server (#14672)
pulumi/docs: content/blog/mcp-server-ai-assistants/index.md

---

> title: "Supercharge Your Infrastructure Workflow: AI-Assisted IaC with Pulumi's Model Context Protocol Server"

Reviewer: Maybe just simplify to:

```suggestion
title: "AI-Assisted Infrastructure as Code with Pulumi's Model Context Protocol Server"
```

social:
    twitter: "Streamline your Pulumi workflow! See how AI assistants + Pulumi's Model Context Protocol (MCP) integration accelerate IaC development by bringing resource discovery, code generation, and validation directly into your editor. #Pulumi #AI #IaC"
    linkedin: "Discover how integrating AI coding assistants (like Cursor, Copilot, etc.) with Pulumi's Model Context Protocol (MCP) server transforms Infrastructure as Code development. This post walks through a real-world example of provisioning an AKS cluster, showcasing reduced context switching, faster discovery, and tighter feedback loops—all within the editor. Boost your team's productivity and improve the developer experience with AI-assisted IaC."
---

Infrastructure as Code (IaC) has revolutionized how we manage cloud resources, but navigating complex cloud provider APIs, writing boilerplate code, and iterating through deployment cycles can still be time-consuming. Pulumi offers a fantastic developer experience using familiar programming languages. But what if we could make it even *faster* and more intuitive by integrating powerful AI assistants directly into the development loop?

<!--more-->

> This is where the **Pulumi Model Context Protocol (MCP) Server integration** shines. By connecting AI-powered code assistants with Pulumi's CLI and registry, we can bring real-time resource information and infrastructure management directly into the development environment, dramatically reducing friction and accelerating workflows.

Reviewer 2: Not sure if we want a tldr about what an MCP is or just a link out to anthropic's docs about them

Several AI coding assistants like GitHub Copilot, Anthropic's Claude Code, Windsurf and others are rapidly evolving; this post will use **Cursor** (an AI-first code editor) to demonstrate a real-world example of this synergy in action.

## Setting up the Pulumi MCP Integration in Cursor

Before diving in, you typically need to configure your AI assistant to communicate with the Pulumi MCP server. In Cursor, this involves:

1. Creating a configuration file named `mcp.json` within the `.cursor` directory in your project's root.

> 2. Populating `mcp.json` with the necessary endpoint and authentication details for the Pulumi MCP server.

Reviewer 2: Can we articulate which JSON key is the endpoint and which is auth?

Author: Removed this line, it's not really relevant

    ```json
    {
        "mcpServers": {
            "pulumi": {
                "type": "stdio",
                "command": "npx",
                "args": ["@pulumi/mcp-server"]
            }
        }
    }
    ```

3. Validating the connection within the assistant's settings (e.g., Cursor has a dedicated section to check the MCP connection status).

    ![Cursor MCP Settings](cursor-mcp-settings.png)

> Once configured, the AI assistant can leverage Pulumi tools seamlessly.

Reviewer 2: Could we shortly include a description of what a tool is

6. The AI assistant processes this output and uses it to continue the task – either by generating code, providing an answer, or deciding on the next step (like suggesting a fix for an error).

This tight loop keeps the developer focused within their editor environment, minimizing disruptions.

![Cursor calling Pulumi tools](cursor-tool-calls.png)

### Step-by-Step Walkthrough

Here’s how the AKS provisioning task played out:

#### The Request

The developer starts with a natural language request to the AI assistant within the editor:

> "I have an empty Pulumi project with TypeScript. Please edit the program to provision an AKS cluster for me. It's a temporary AKS cluster that I need for a short experiment, so I don't need any particular configuration of it. Just export its kubeconfig when you are done. Please use the tools to lookup resource information and to run pulumi preview to make sure your code works. When done, run pulumi up for me."

#### AI-Powered Resource Discovery

Instead of the developer manually searching docs, the AI assistant leverages the MCP integration.

> - It first queries the Pulumi Registry via MCP to list resources within the Azure Native provider's `containerservice` module (using the `pulumi_registry_listResources` tool). This immediately identified `ManagedCluster` as the relevant resource.

Reviewer 2: ```suggestion
- The AI assistant first queries the Pulumi Registry via MCP to list resources within the Azure Native provider's `containerservice` module (using the `pulumi_registry_listResources` tool). This immediately identified `ManagedCluster` as the relevant resource.
```

> - It then requested detailed information *specifically* for `ManagedCluster` (using `pulumi_registry_getResource`), directly retrieving its required properties, descriptions, and structure without leaving the editor.

Reviewer 2: ```suggestion
- The AI assistant then requests detailed information *specifically* for `ManagedCluster` (using `pulumi_registry_getResource`), directly retrieving its required properties, descriptions, and structure without leaving the editor.
```
